openverse_catalog/dags/providers/provider_api_scripts/nypl.py

```diff
@@ -1,211 +1,248 @@
+        if not isinstance(creatorinfo, list):
+            creatorinfo = [creatorinfo]
         primary_creator = (
             info.get("namePart", {}).get("$")
             for info in creatorinfo
             if info.get("usage") == "primary"
         )
         creator = next(primary_creator, None)
-    else:
-        creator = None
-
-    if creator is None:
-        logger.warning("No primary creator found")
-
-    return creator
-
-
-def _get_image_url(images, image_url_dimensions=None):
-    if image_url_dimensions is None:
-        image_url_dimensions = IMAGE_URL_DIMENSIONS
-    image_type = {
-        parse_qs(urlparse(img.get("$")).query)["t"][0]: img.get("$") for img in images
-    }
-
-    image_url = _get_preferred_image(image_type, image_url_dimensions)
-
-    return image_url
 
+        return creator
+
+    @staticmethod
+    def _get_metadata(mods):
+        metadata = {}
+
+        type_of_resource = mods.get("typeOfResource")
+        if isinstance(type_of_resource, list) and (
+            type_of_resource[0].get("usage") == "primary"
+        ):
+            metadata["type_of_resource"] = type_of_resource[0].get("$")
+
+        if isinstance(mods.get("genre"), dict):
+            metadata["genre"] = mods.get("genre").get("$")
+
+        origin_info = mods.get("originInfo", {})
+        if date_issued := origin_info.get("dateIssued", {}).get("$"):
+            metadata["date_issued"] = date_issued
+        if date_created := origin_info.get("dateCreated", {}):
+            if isinstance(date_created, list):
+                # Approximate dates have a start and an end
+                # [{'encoding': 'w3cdtf', 'keyDate': 'yes',
+                # 'point': 'start', 'qualifier': 'approximate', '$': '1990'},
+                # {'encoding': 'w3cdtf', 'point': 'end',
+                # 'qualifier': 'approximate', '$': '1999'}]
+                start_date = next(
+                    d.get("$") for d in date_created if d.get("point") == "start"
+                )
+                end_date = next(
+                    d.get("$") for d in date_created if d.get("point") == "end"
+                )
+                if start_date and end_date:
+                    date_created = f"{start_date}-{end_date}"
+                elif start_date:
+                    date_created = start_date
+                else:
+                    date_created = None
+            else:
+                date_created = date_created.get("$")
+            metadata["date_created"] = date_created
+        if publisher := origin_info.get("publisher", {}).get("$"):
+            metadata["publisher"] = publisher
+
+        if description := mods.get("physicalDescription", {}).get("note", {}).get("$"):
+            metadata["physical_description"] = description
+
+        subject_list = mods.get("subject", [])
+        if isinstance(subject_list, dict):
+            subject_list = [subject_list]
+        # Topic can be a dictionary or a list
```

Comment:
How annoying!

Reply:
It is everywhere in the API replies :(
I've added a helper to simplify some usecases.